server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java

```diff
@@ -0,0 +1,226 @@
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(HeaderMatcher.class);
+
+        private final ContentMatcher contentMatcher;
+        private final String ruleValue;
+        private final HeaderExtractor headerExtractor;
+
+        HeaderMatcher(ContentMatcher contentMatcher, String ruleValue,
+                      HeaderExtractor headerExtractor) {
+            Preconditions.checkNotNull(contentMatcher);
+            Preconditions.checkNotNull(headerExtractor);
+
+            this.contentMatcher = contentMatcher;
+            this.ruleValue = ruleValue;
+            this.headerExtractor = headerExtractor;
+        }
+
+        @Override
+        public boolean match(Mail mail) {
+            try {
+                Stream<String> headerLines = headerExtractor.apply(mail);
+                return contentMatcher.match(headerLines, ruleValue);
+            } catch (Exception e) {
+                LOGGER.error("error while extracting mail header", e);
+                return false;
+            }
+        }
+    }
+
+    interface ContentMatcher {
+
+        class AddressHeader {
+            private static final Logger LOGGER = LoggerFactory.getLogger(AddressHeader.class);
+
+            private Optional<String> personal;
+            private Optional<String> address;
+            private final String fullAddress;
+
+            private AddressHeader(String fullAddress) {
+                this.fullAddress = fullAddress;
+                parseFullAddress();
+            }
+
+            private void parseFullAddress() {
+                try {
+                    InternetAddress internetAddress = new InternetAddress(fullAddress);
+                    this.personal = Optional.ofNullable(internetAddress.getPersonal());
+                    this.address = Optional.ofNullable(internetAddress.getAddress());
+                } catch (AddressException e) {
+                    LOGGER.error("error while parsing full address {}", fullAddress, e);
+                    this.personal = Optional.empty();
+                    this.address = Optional.empty();
+                }
+            }
+
+            public Optional<String> getPersonal() {
+                return personal;
+            }
+
+            public Optional<String> getAddress() {
+                return address;
+            }
+
+            public String getFullAddress() {
+                return fullAddress;
+            }
+        }
+
+        ContentMatcher STRING_CONTAINS_MATCHER = (contents, valueToMatch) -> contents.anyMatch(content -> StringUtils.contains(content, valueToMatch));
+        ContentMatcher STRING_NOT_CONTAINS_MATCHER = negate(STRING_CONTAINS_MATCHER);
+        ContentMatcher STRING_EXACTLY_EQUALS_MATCHER = (contents, valueToMatch) -> contents.anyMatch(content -> StringUtils.equals(content, valueToMatch));
+        ContentMatcher STRING_NOT_EXACTLY_EQUALS_MATCHER = negate(STRING_EXACTLY_EQUALS_MATCHER);
```

Comment:
IMO we don't need that as it is not API exposed.

It would also beat back (cardinality is already a pain here)

However I 100% agree with the `and maybe address matcher should be case insensitive by default?` claim. We should add a test for this, and support this behaviour.